### PR TITLE
PubMatic: add "acat" parameter

### PR DESCRIPTION
**Pull request by pm-nilesh-chate**
Add support to 'act' request ext param in PubMatic adapter.

Documentation https://docs.prebid.org/dev-docs/pbs-bidders.html#pubmatic (https://github.com/prebid/prebid.github.io/pull/3657)

PBJS: https://github.com/prebid/Prebid.js/pull/8246

### CODE REVIEW

File: adapters/pubmatic/pubmatic.go [hunk 1]
```diff
@@ -310,23 +315,29 @@ func parseImpressionObject(imp *openrtb2.Imp, extractWrapperExtFromImp, extractP
...
-func extractPubmaticWrapperExtFromRequest(request *openrtb2.BidRequest) (*pubmaticWrapperExt, error) {
-	var wrpExt pubmaticWrapperExt
+// extractPubmaticExtFromRequest parse the req.ext to fetch wrapper and acat params
+func extractPubmaticExtFromRequest(request *openrtb2.BidRequest) (*pubmaticWrapperExt, []string, error) {
+	var acat []string
+	var wrpExt *pubmaticWrapperExt
 	reqExtBidderParams, err := adapters.ExtractReqExtBidderParamsMap(request)
 	if err != nil {
-		return nil, err
+		return wrpExt, acat, err
```
Comment (SyntaxNode) [inline adapters/pubmatic/pubmatic.go:324]: Nitpick: IMHO it's easier to keep the `nil`s here to make it clear there are no values.

Comment (pm-nilesh-chate) [reply adapters/pubmatic/pubmatic.go]: done

**Comment by pm-nilesh-chate** [comment]
> Please update `static/bidder-params/pubmatic.json` to include the new acat bidder parameter.

done

File: adapters/pubmatic/pubmatic.go [hunk 2]
```diff
@@ -310,23 +315,32 @@ func parseImpressionObject(imp *openrtb2.Imp, extractWrapperExtFromImp, extractP
...
-		err = json.Unmarshal(wrapperObj, &wrpExt)
+		wrpExt = &pubmaticWrapperExt{}
+		err = json.Unmarshal(wrapperObj, wrpExt)
 		if err != nil {
-			return nil, err
+			return nil, acat, err
+		}
+	}
+
+	if _acat, ok := reqExtBidderParams["acat"]; ok {
```
Comment (AlexBVolcy) [inline adapters/pubmatic/pubmatic.go:336]: Could you explain why there's an underscore in front of acat here `_acat`

File: adapters/pubmatic/pubmatic.go [hunk 3]
```diff
@@ -66,7 +66,7 @@ func (a *PubmaticAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *ad
 	extractWrapperExtFromImp := true
 	extractPubIDFromImp := true
 
-	wrapperExt, err := extractPubmaticWrapperExtFromRequest(request)
+	wrapperExt, acat, err := extractPubmaticExtFromRequest(request)
 	if err != nil {
 		return nil, []error{err}
```
Comment (AlexBVolcy) [inline adapters/pubmatic/pubmatic.go:71]: Could you add a test to cover the error case?

File: adapters/pubmatic/pubmatic.go [hunk 4]
```diff
@@ -114,9 +114,14 @@ func (a *PubmaticAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *ad
...
+		reqExt["acat"] = acat
+	}
 	if wrapperExt != nil {
-		reqExt := make(map[string]interface{})
 		reqExt["wrapper"] = wrapperExt
+	}
+	if len(reqExt) > 0 {
 		rawExt, err := json.Marshal(reqExt)
 		if err != nil {
 			return nil, []error{err}
```
Comment (AlexBVolcy) [inline adapters/pubmatic/pubmatic.go:127]: Could you add a test to cover this error case

**Review by AlexBVolcy — COMMENTED** [review]
Looks good, left a few small comments

File: adapters/pubmatic/pubmatic.go (hunk 2 above)
Comment (pm-nilesh-chate) [reply adapters/pubmatic/pubmatic.go:336]: Was temporary variable. Renamed it to `acatBytes`

File: adapters/pubmatic/pubmatic.go (hunk 3 above)
Comment (pm-nilesh-chate) [reply adapters/pubmatic/pubmatic.go:71]: done

File: adapters/pubmatic/pubmatic.go (hunk 4 above)
Comment (pm-nilesh-chate) [reply adapters/pubmatic/pubmatic.go:127]: This is a bit tricky to cover under unit test as the marshalled instance is initiated locally and would always have the correct instance in it.

**Review by AlexBVolcy — DISMISSED** [review]
LGTM

File: adapters/pubmatic/pubmatictest/exemplary/banner.json [hunk 5]
```diff
@@ -33,6 +33,13 @@
         "device":{
             "ua": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.132 Safari/537.36"
         },
+        "ext": {
+          "prebid": {
+            "bidderparams": {
+              "acat": ["drg”,”dlu”,”ssr"]
```
Comment (SyntaxNode) [inline adapters/pubmatic/pubmatictest/exemplary/banner.json:39]: I don't think this is testing what you think its testing. The inner double quotes are actually a different character such that this (and the same in other files) is an array of 1 item. Is that intentional?

Comment (pm-nilesh-chate) [reply adapters/pubmatic/pubmatictest/exemplary/banner.json]: Fixed. Not sure how the quotes were changed. I doubt it was json validator plugin.
Thanks for pointing it out :)

**Review by SyntaxNode — APPROVED** [review]

**Review by AlexBVolcy — APPROVED** [review]

**Pull request merged**